### PR TITLE
Create proposal and vote

### PR DESCRIPTION
**Pull request by olgaklimenko**

### CODE REVIEW

File: src/MintingPolicy/Governance.hs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+governancePolicy = plam $ \ref rdm' ctx -> P.do
+  (rdm, _) <- ptryFrom @GovernanceTokensRedeemer rdm'
+  txInfo <- plet $ pfield @"txInfo" # ctx
+  pmatch rdm $ \case
+    PMintGovernanceTokens mintFields -> popaque $
+      unTermCont $ do
+        amount <- pletC $ pfield @"_0" # mintFields
+        red <- pletFieldsC @["_0", "_1"] mintFields
+        checkUTxOSpent ref ctx
+        checkMintingAmount red._0 govTokenName ctx
```
Comment (olgaklimenko) [inline src/MintingPolicy/Governance.hs:40]: remove unnecessary checks

File: src/MintingPolicy/Proposal.hs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    amountBs <- plet $ pencodeUtf8 # (pshow amount)
+    separator <- plet $ pconsBS # 46 # mempty -- 46 is ascii for "."
+    tnBs <- plet $ headV <> voteBs <> separator <> amountBs
+    plet tnBs $ \tnBs ->
+      pif
+        (plengthBS # tnBs #<= tokenNameSizeLimit)
+        (pcon . PTokenName $ tnBs)
+        (ptraceError "601")
+
+-- parseVoteTokenName :: Term s (PByteString :--> PTuple PInteger PInteger)  -- Result: (Vote, Amount)
```
Comment (olgaklimenko) [inline src/MintingPolicy/Proposal.hs:90]: rm

File: src/Governance/Proposal/Validator.hs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+checkPaymentToVoter :: 
+  Term s PPubKeyHash
+  -> Term s PInteger
+  -> Term s PInteger
+  -> Term s PCurrencySymbol
+  -> Term s (PAsData PTxInfo)
+  -> TermCont s ()
+checkPaymentToVoter pkh vote amount proposalCs txInfo = do
+  voteTokenName <- pletC $ makeVoteTn # vote # amount
```
Comment (olgaklimenko) [inline src/Governance/Proposal/Validator.hs:105]: check that vote token minted here

File: src/Governance/Proposal/Validator.hs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            voterPkh <- pletC $ extractPaymentPkhFromAddress # redData._2
+            checkIsSignedBy "903" voterPkh txInfo
+            pguardC "912" $ inDatum.applied #== pdata 0
+            checkVotedBeforeDeadline inDatum.deadline txInfo
+
+            pguardC "904" $ inDatum.proposal #== outDatum.proposal
+            pguardC "905" $ inDatum.policyRef #== outDatum.policyRef
+            pguardC "906" $ inDatum.quorum #== outDatum.quorum
+            pguardC "907" $ inDatum.initiator #== outDatum.initiator 
+            pguardC "908" $ inDatum.deadline #== outDatum.deadline --
```
Comment (olgaklimenko) [inline src/Governance/Proposal/Validator.hs:57]: remove comment

File: src/Governance/Proposal/Validator.hs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  -> Term s PInteger
+  -> Term s PScriptContext
+  -> TermCont s ()
+checkOutputValue inValue' outValue' protocolCs amount ctx = do
+  inValue <- pletC $ Value.pforgetPositive inValue'
+  outValue <- pletC $ Value.pforgetPositive outValue'
+  govDatum' <- pletC $ getGovernanceDatumFromReferenceUtxo # protocolCs # ctx
+  govDatum <- pletFieldsC @["govCurrency", "govTokenName"] govDatum'
+  expectedAdaoAdding <- pletC $ Value.psingleton # govDatum.govCurrency # govDatum.govTokenName # amount
+  expectedOutputValue <- pletC $ inValue <> expectedAdaoAdding <> minAdaValue
```
Comment (olgaklimenko) [inline src/Governance/Proposal/Validator.hs:94]: rm Adao

File: src/MintingPolicy/Proposal.hs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      unTermCont $ do
+        let tokenName = pfield @"_0" # burnFields
+        checkMintingAmount (-1) tokenName ctx
+        pure $ pconstant ()
+    PMintVoteToken mintFields' -> popaque $
+      unTermCont $ do
+        mintFields <- pletFieldsC @["_0", "_1", "_2"] mintFields'
+        voteTokenName <- pletC $ makeVoteTn # mintFields._1 # mintFields._2
+        checkMintingAmount 1 voteTokenName ctx
+        checkProposalTokensInInput mintFields._0 ctx
```
Comment (olgaklimenko) [inline src/MintingPolicy/Proposal.hs:56]: remove all checks except `checkProposalTokensInInput`

File: src/Governance/Proposal/Datum.hs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+          ( PDataRecord
+              '[ "minAmount" ':= PInteger
+               , "maxAmount" ':= PInteger
+               , "minDuration" ':= PInteger
+               , "maxDuration" ':= PInteger
+               , "protocolFee" ':= PInteger
+               ]
+          )
+      )
+
```
Comment (KateBushueva) [inline src/Governance/Proposal/Datum.hs:52]: I think we can combine it with the changable Protocol Parameters somehow later to have only one structure.. maybe leave a TODO here?

**Review by KateBushueva — APPROVED** [review]

**Pull request merged**